compiler/rustc_hir_analysis/src/astconv/generics.rs

```diff
@@ -369,9 +378,22 @@ pub fn create_args_for_parent_generic_args<'tcx, 'a>(
                     //     user.
                     //  3. We've inferred some lifetimes, which have been provided later (i.e.
                     //     after a type or const). We want to throw an error in this case.
+                    //  4. A host effect argument was passed but the generic params don't
+                    //     contain a host effect param. This can happen if a `~const` trait
+                    //     bound modifier was placed on a non-const trait. Since the lowerer
+                    //     has no way of knowing if the trait is const or not, it synthesizes
+                    //     host effect args unconditionally. We will emit an error later and
+                    //     thus we can ignore it.
 
                     if arg_count.correct.is_ok()
                         && arg_count.explicit_late_bound == ExplicitLateBound::No
+                        && !matches!(
+                            arg,
+                            GenericArg::Const(hir::ConstArg {
+                                is_desugared_from_effects: true,
+                                ..
+                            })
+                        )
```

Comment:
Obviously I'm “treating symptoms” here but there's no way for the lowering step to know this information up front before name resolution.